game_of_life.rb

```diff
@@ -1,133 +1,56 @@
-# XXX: OVERVIEW
-# In this solution, state is stored in a hash where each hash key is an integer
-# and represents each row in the 2D grid. Each "row" (hash[key]) is an array of
-# integers. Each integer in the "row" array, represents a cell's location.
-# The cells that exist in a given row are implicitly alive.
-# Non-existent cells are implicitly dead. Dead cells have no bearing on the
-# fate of neighboring cells and therefore are not tracked.
-
-# XXX: EXAMPLE STATE
-# {
-#   1 => [1, 2,    4],
-#   2 => [   2      ],
-#   3 => [1, 2, 3, 4]
-# }
-
-# XXX: PROJECTIONS
-# subsequent states are calculated by projecting scores for the possible cells
-# that surround only the living (existing) cells.
-
-# XXX: EXAMPLE PROJECTION
-# {
-#   # row
-#   1 => {
-#     # cells and their respective scores
-#     1 => 4,
-#     3 => 5
-#   },
-#
-#   # row
-#   2 => {
-#     1 => 2,
-#     3 => 4
-#   }
-# }
-
+require_relative 'cell'
 
 class GameOfLife
 
   attr_reader :state
 
-  def initialize
-    @state = {}
-  end
-
-  def seed(size = 10)
-    for i in (1..size) do
-      @state[i] = random_row(size)
-    end
+  def initialize(initial_state)
+    @state = []
+    @future = []
+    initial_state.each { |coords| @state << Cell.new(coords) }
+    build_future
   end
 
-  def set_state(new_state)
-    @state = new_state
-  end
-
-  def increment(projection, y, x)
-    projection[y] ||= {}
-    projection[y][x] ||= 0
-    projection[y][x] += 1
-  end
-
-  def project_row(row_num, row_cells)
-    projection = {}
-    row_cells.each do |cell|
-      for i in (-1..1) do
-        for j in (-1..1) do
-          increment(projection, row_num+i, cell+j) unless (i == 0 && j == 0)
-        end
+  def build_future
+    @future.clear
+    state.each do |cell|
+      cell.influence.each do |neighbor|
```

Comment:
@cromwellryan the game doesn't necessarily have a grid... just coords. The grid would be established by a consumer. Or at least that's how I envisioned it.

Also, zero living cells would never produce anything so... I guess the game wouldn't exist.